### PR TITLE
Invoices send instantly

### PR DESCRIPTION
**Pull request by tmb**
This PR forces invoices to immediately finalize and send out (vs. waiting for Stripe to do it for us)

In order for it to work, we'll need to tweak Stripe (found under Billing Settings):
screen shot 2018-12-30 at 3 30 03 pm

Also- currently, sponsors are never sent reminders after they receive an invoice, I feel like they should be reminded after like 7 business days, thoughts?

### CODE REVIEW

File: app/models/invoice.rb [hunk 1]
```diff
@@ -169,7 +169,11 @@ def create_stripe_invoice
     inv = StripeService::Invoice.create(stripe_invoice_params)
     self.stripe_invoice_id = inv.id
 
+    inv = inv.finalize_invoice
```
Comment (maxwofford) [inline app/models/invoice.rb]: Might get confusing to read this code later with redeclaring this.

Thoughts on calling this `finalized_inv`?
```rb
finalized_inv = inv.finalize_invoice
self.set_fields_from_stripe_invoice(finalized_inv)
finalized_inv.send_invoice
```

or making it clear the initial declaration is a draft?
```rb
draft = StripeService::Invoice.create(stripe_invoice_params)
self.stripe_invoice_id = draft.id
inv = draft.finalize_invoice
```

Comment (tmb) [reply app/models/invoice.rb]: agree that showing the declaration as a draft makes more sense here, will fix.

**Comment by tmb** [comment]
@MaxWofford updated, can you review?

File: app/models/invoice.rb [hunk 2]
```diff
@@ -166,10 +166,14 @@ def create_stripe_invoice
     item = StripeService::InvoiceItem.create(stripe_invoice_item_params)
     self.item_stripe_id = item.id
 
-    inv = StripeService::Invoice.create(stripe_invoice_params)
-    self.stripe_invoice_id = inv.id
+    draft = StripeService::Invoice.create(stripe_invoice_params)
+    self.stripe_invoice_id = draft.id
+
+    inv = draft.finalize_invoice
```
Comment (zachlatta) [inline app/models/invoice.rb]: Have you tested this? I’m a bit worried that we might need to upgrade the API version we’re using, which will break other things, to use the new `/send` endpoint on invoices (that’s what I’m assuming `finalize_invoice` is using under the hood).

File: app/models/invoice.rb (hunk 1 above)
Comment (zachlatta) [reply app/models/invoice.rb]: Are you sure `finalize_invoice` doesn't just modify the object directly? IIRC Stripe's Ruby API has mutable objects with methods that change underlying objects directly.

Comment (tmb) [reply app/models/invoice.rb]: Pretty sure but might be worth trying to see if it does mutate, their docs show that it doesn't

File: app/models/invoice.rb (hunk 2 above)
Comment (tmb) [reply app/models/invoice.rb]: yup, didn't test this or realize that we were on the older API version. maybe we can push this into 1.0 vs. trying to do it now?

Comment (tmb) [reply app/models/invoice.rb]: hmm stripe's changelogs don't show that the /finalize endpoint was added/change via https://stripe.com/docs/upgrades

**Comment by tmb** [comment]
Going to go ahead and close this + do a lot more testing (probably bump the Stripe API version and push the upgraded working stuff to `one`) because this definitely can't just go right on the master branch.

**Comment by zachlatta** [comment]
👍

--
Zach Latta

https://zachlatta.com ( http://zachlatta.com )
@zachlatta ( https://twitter.com/zachlatta ) / fb ( https://www.facebook.com/crynix ) / github ( https://github.com/zachlatta )

**Pull request closed without merging**